p2panda-net/src/sync/manager.rs

```diff
@@ -322,64 +303,53 @@ where
+        if let Some(attempt) = self.sessions.get_mut(&scope) {
+            attempt.status = Status::Complete(Instant::now())
         }
 
-        if let Some(err) = err.downcast_ref::<SyncAttemptError>() {
-            // If the sync attempt failed for any reason we want to retry up to
-            // `max_retry_attempts`.
-            warn!("sync attempt failed: {err}");
-
-            // We need to inform the engine of the failed attempt so that the gossip buffer counter
-            // can be decremented.
-            self.engine_actor_tx
-                .send(ToEngineActor::SyncFailed {
-                    topic: Some(sync_attempt.topic.clone()),
-                    peer: sync_attempt.peer,
-                })
-                .await?;
-
-            if sync_attempt.attempts <= self.config.max_retry_attempts {
-                self.reschedule_attempt(sync_attempt).await?;
-                return Ok(());
-            }
+        if self.config.is_resync() {
+            self.resync_queue.push_back(scope);
         }
 
-        // @TODO(glyph): We may want to maintain a map of failed peer-topic combinations that can
-        // be checked against each announcement received by the sync manager. Otherwise we may run
-        // into the case where we are repeatedly initiating a sync session with a faulty peer (this
-        // would happen every time we receive an announcement, approximately every 2.2 seconds).
-
         Ok(())
     }
 
-    /// Remove the given topic from the set of active sync sessions for the given peer and add them
-    /// to the set of completed sync sessions.
+    /// Mark the status of the attempt as `Failed`, increment the attempts counter and inform the
+    /// engine of the failure.
     ///
-    /// If resync is active, a timestamp is created to mark the time of sync completion and the
-    /// attempt is then pushed to the back of the resync queue.
-    async fn complete_successful_sync(&mut self, mut sync_attempt: SyncAttempt<T>) -> Result<()> {
-        trace!("complete_successful_sync");
-        self.completed_sync_sessions
-            .entry(sync_attempt.topic.clone())
-            .or_default()
-            .insert(sync_attempt.peer);
-
-        if let Some(session) = self.active_sync_sessions.get_mut(&sync_attempt.topic) {
-            session.remove(&sync_attempt.peer);
-        }
-
-        if self.config.is_resync() {
-            trace!("schedule re-sync attempt");
-            sync_attempt.completed = Some(Instant::now());
-            self.resync_queue.push_back(sync_attempt);
+    /// The attempt is pushed to the back of the retry queue if the maximum number of retry
+    /// attempts has not been exceeded.
+    async fn complete_failed_sync(&mut self, scope: Scope<T>, err: Error) -> Result<()> {
+        warn!("sync attempt failed for scope {:?}: {}", scope, err);
+
+        // @TODO(glyph): There is a bug in this flow. If the sync initiation attempt fails before
```

Comment:
Here I describe the gossip buffer issue. We should be sending `SyncFailed` but it causes the `retry_after_error` test to fail because the gossip buffer unlocking code panics. I propose we rather not panic and simply log a warning.

Reply:
Just to check I understand: if a sync session is initiated but then fails before `HandshakeSuccess` message is received, we never locked our gossip buffer, and so when we handle the `SyncFailed` event and decrement the buffer a panic occurs in the engine because the buffer counter can go below zero. It could be that there are other sync sessions occurring, and so it wouldn't panic yet, but it _would_ panic later when a different sync session finished and tried to decrement the same buffer counter. 

I can't think of any downside to removing the panic in the engine. In all cases it seems that there are no sync sessions running and so the buffer should be unlocked.

Reply:
Your understanding is exactly correct.